transport/src/behaviour/request_client.rs

```diff
@@ -165,22 +169,40 @@ where
         }))
     }
 
-    fn on_failure(&mut self, peer_id: PeerId, req_id: OutboundRequestId) -> Option<TToSwarm<Self>> {
-        log::debug!("Request {req_id} failed");
-        // If request was already resubmitted, remove it and emit timeout event
-        if let Some(req_id) = self.resubmitted_requests.remove(&req_id) {
-            return Some(ToSwarm::GenerateEvent(ClientEvent::Timeout { peer_id, req_id }));
-        }
+    fn on_failure(
+        &mut self,
+        peer_id: PeerId,
+        mut req_id: OutboundRequestId,
+        error: OutboundFailure,
+    ) -> Option<TToSwarm<Self>> {
+        log::debug!("Request {req_id} failed: {error}");
 
-        // If request was submitted for the first time, try to find peer and connect
-        self.waiting_for_connection.entry(peer_id).or_default().insert(req_id);
-        if !self.lookup_timeouts.contains(peer_id) {
-            log::debug!("Requesting lookup for peer {peer_id}");
-            _ = self.lookup_timeouts.try_push(peer_id, futures::future::pending());
-            return Some(ToSwarm::GenerateEvent(ClientEvent::PeerUnknown { peer_id }));
+        // If request was submitted for the first time and dial failed, try to find peer and connect
+        if matches!(&error, OutboundFailure::DialFailure)
```

Comment:
Could you help to understand what changed here?

Reply:
Sure. Previous logic was:
```
1. If the request has been re-submitted, and failed, return timeout.
2. Otherwise:
  2.1 Put the request in a queue to resend when a connection is established, and
  2.2 If there is no lookup for the peer going on, pass an event indicating that we need to find and dial the peer.
```
It didn't differentiate the original error, which was a mistake.

The current logic is:
```
1. If the request hasn't been re-submitted yet, and the original error was DialFailure (which might indicate that we don't know a dialable address for the peer):
  1.1 Put the request in a queue to be resent when a connection is established, and
  1.2 If there is no active lookup, start a lookup and dial.
2. Otherwise:
  2.1 If the original error was Timeout return Timeout,
  2.2 Otherwise return Failure.
```
The main change is that we attempt to find the node and dial only on `DialFailure`. In [other cases](https://docs.rs/libp2p/latest/libp2p/request_response/enum.OutboundFailure.html) it doesn't make sense:
* `Timeout` – The request has timed out already, we have no time to do the lookup.
* `ConnectionClosed` – There was already a connection to the peer. Perhaps the peer just has a bad network connection or it crashed.
* `UnsupportedProtocols` – The peer doesn't even support executing queries.
* `Io` – Similar as `ConnectionClosed`, the issue is not that we don't know the peer address.